docs/x509/verification.rst

```diff
@@ -297,3 +323,182 @@ the root of trust:
+
+    .. method:: maybe_present(oid, criticality, validator_cb)
+
+        Specifies that the extension identified by the given OID may or may not be present.
+        If it is present, it must conform to the given criticality constraint. 
+        An optional validator callback may be provided.
+
+        If a validator callback is provided, the callback will be invoked 
+        when :meth:`ClientVerifier.verify` or :meth:`ServerVerifier.verify` is called on a verifier 
+        that uses the extension policy. For details on the callback signature, see :type:`MaybeExtensionValidatorCallback`.
+
+        :param ObjectIdentifier oid: The OID of the extension that may be present
+        :param Criticality criticality: The criticality of the extension
+        :param validator_cb: An optional Python callback to validate the extension value. 
+        :type validator_cb: :type:`MaybeExtensionValidatorCallback` or None
+
+        :returns: An instance of :class:`ExtensionPolicyBuilder`
+
+    .. method:: must_be_present(oid, criticality, validator_cb)
+
+        Specifies that the extension identified by the given OID must be present and conform to the given criticality constraint.
+        An optional validator callback may be provided.
+
+        If a validator callback is provided, the callback will be invoked 
+        when :meth:`ClientVerifier.verify` or :meth:`ServerVerifier.verify` is called on a verifier 
+        that uses the extension policy. For details on the callback signature, see :type:`PresentExtensionValidatorCallback`.
+
+        :param ObjectIdentifier oid: The OID of the extension that must be present
+        :param Criticality criticality: The criticality of the extension
+        :param validator_cb: An optional Python callback to validate the extension
+        :type validator_cb: :type:`PresentExtensionValidatorCallback` or None
+
+        :returns: An instance of :class:`ExtensionPolicyBuilder`
+
+    .. method:: build()
+
+        Builds the extension policy.
+
+        :returns: An :class:`ExtensionPolicy`
+
+.. class:: Criticality
+
+    .. versionadded:: 44.0.0
+
+    An enumeration of criticality constraints for certificate extensions.
+
+    .. attribute:: CRITICAL
+
+        The extension must be marked as critical.
+
+    .. attribute:: AGNOSTIC
+            
+        The extension may be marked either as critical or non-critical.
+
+    .. attribute:: NON_CRITICAL
+
+        The extension must not be marked as critical.
+
+.. class:: ExtensionPolicy
+
+    .. versionadded:: 44.0.0
+
+    An ExtensionPolicy constrains the presence, contents and criticalities of certificate extensions.
+
+    This type is opaque to the user and should be created using :class:`ExtensionPolicyBuilder`.
+    Pass the created policy to :meth:`PolicyBuilder.ca_extension_policy` or :meth:`PolicyBuilder.ee_extension_policy`
+    to set the policies used during verification.
+
+.. class:: Policy
```

Comment:
Hmm, this feels duplicative of our existing types, which makes me feel that something is off. But I don't have a concrete suggestion.

Reply:
Yes, this does indeed closely duplicate the rust `Policy` struct in `cryptography-x509-verification`.  I couldn't really come up with any way to avoid this since that module is supposed to be completely python-agnostic, and we need to pass the policy parameters to the python extension validator callback. 

I definitely welcome any suggestions on how to make this better, this is definitely something I would like to avoid, but I think it might not be possible without dropping the constraint of `cryptography-x509-verification` being python-agnostic.

Reply:
@woodruffw do you have any name or API suggestions here?

Reply:
The only thing that comes to mind in terms of naming changes is something like `ExtensionValidationContext`/`ExtensionValidatorContext`, but I don't know if that makes things better. Throwing it out there just in case

Reply:
(Currently the only use of this class is passing it to user-provided extension validators)

Reply:
My current best idea:

`PolicyBuilder().build_server_verifier()` -> `ServerVerifier`

`ServerVerifier.policy` -> `Policy`

`subject`, `validation_time`, `store`, `max_chain_depth`, which are currently attributes on `ServerVerifier` become attributes on `Policy`.

(Same for `ClientVerifier` obviously.)

WDYT?

Reply:
Yes, that definitely sounds like an improvement. I didn't even notice this aspect of the duplication - that the fields on verifiers mirror Policy fields. Honestly, this change would kinda compensate for the added duplication between the rust-only and the python-exposed `Policy` structs by reducing existing duplication between the two Verifiers which had basically the exact same getters.

Reply:
I still don't think that adding a "python version" of `cryptography-x509-verification::Policy` is avoidable though. I think that might unfortunately be a pattern that we just have to live with. It already naturally occurs in other places, e.g. `Certificate`, although the duplication there feels more warranted with addition of logic that is specific to the "python wrapper"

Reply:
(It might be possible to change with some global refactor ofc, but I'm not considering anything as drastic 😄)

Reply:
So, are we rolling with your idea and saying that that's enough, or do we still want to look for other solutions?
I am would be content with implementing your idea tbh, I think that would be enough. (What I mean is I don't have any better ideas 😄)

Reply:
I don't have any better ideas :-)

Reply:
I edited the docs and implemented this in my local branch where I'm preparing the actual code for the feature, so I'm resolving this for now. Feel free to un-resolve if you have any other comments regarding this.